system/jlib/jstring.cpp

```diff
@@ -937,27 +937,58 @@ StringBuffer & StringBuffer::replace(char oldChar, char newChar)
 // Copy source to result, replacing all occurrences of "oldStr" with "newStr"
 StringBuffer &replaceString(StringBuffer & result, size_t lenSource, const char *source, size_t lenOldStr, const char* oldStr, size_t lenNewStr, const char* newStr)
 {
-    if (lenSource)
+    // Only copy if the source and search (oldStr) strings are non-empty
+    if (lenSource && lenOldStr)
     {
+        // Scan for match on first character
+        size_t offset = 0;
         size_t left = lenSource;
         while (left >= lenOldStr)
         {
-            if (memcmp(source, oldStr, lenOldStr)==0)
+            if (source[offset] != oldStr[0])
             {
+                offset++;
+                left--;
+            }
+            else
+                break;
+        }
+
+        // Reserve space when the new string is longer than the old string
+        unsigned steps = lenSource - lenOldStr - offset + 1;
+        unsigned maxResultLength = lenNewStr > lenOldStr ? lenSource + steps * (lenNewStr - lenOldStr) : lenSource;
+        result.ensureCapacity(maxResultLength);
+
+        // Add any characters before the match
+        result.append(offset, source);
+
+        size_t unmatchedChars = 0;
+        while (left >= lenOldStr)
+        {
+            if (memcmp(source + offset, oldStr, lenOldStr)==0)
+            {
+                if (unmatchedChars)
+                {
+                    // Copy mismatched characters in blocks
+                    result.append(unmatchedChars, source + offset - unmatchedChars);
+                    unmatchedChars = 0;
+                }
                 result.append(lenNewStr, newStr);
-                source += lenOldStr;
+                offset += lenOldStr;
                 left -= lenOldStr;
             }
             else
             {
-                result.append(*source);
-                source++;
+                unmatchedChars++;
+                offset++;
```

Comment:
Can this else clause repeat the logic of the first while() loop in the function, to compare single characters and increment unmatchedChars without going through memcmp each time?

Reply:
Yes, I have changed it so it compares single characters until it finds a match to the first character.